---
title: "Lab 13 in-class"
output:
  html_document:
    df_print: paged
  html_notebook: default
---

### DNAStrings
```{r}
library(Biostrings)
library(rtracklayer)
library(tidyverse)
```

```{r}
# Create a DNA string
dna.seq <- DNAString("ACGCTTA")
dna.seq
```

```{r}
# Create a set of DNA string
dna.set <- DNAStringSet(c("ACG","GGG","CTCAGGAGTT"))
dna.set
```

```{r}
width(dna.set) # Gives the length of each string
length(dna.set) # Gives the number of strings
```

```{r}
# Do not use tidyverse with biostrings
dna.set %>%
  as.data.frame()
```

```{r}
# DNAStringSets work like a list()
dna.set[[1]]
```
```{r}
dna.set[1]

```

```{r}
# Read in yeast genome
yeast.fasta <- readDNAStringSet("~/Fall2019/Lab-13/yeast.fasta")
yeast.fasta

# Read in yeast GRanges
yeast.gff <- readGFFAsGRanges("~/Fall2019/Lab-13/saccharomyces_cerevisiae.gff.txt")
```

```{r}
yeast.gff
```

```{r}
# Subsetting a genome seq to get a seq of a particular range
telomere.seq <- yeast.fasta[yeast.gff[2]]
telomere.seq
```

```{r}
# Extract chromosome 15
yeast.xv <- yeast.fasta[names(yeast.fasta) == "chrXV"] # This is still a DNAStringSet

# Extract the DNAString from the string set using list notation
yeast.xv.dnastring <- yeast.xv[[1]]

# Extract bases at positions 100 to 105
yeast.xv.dnastring[100:105]
```
 
### Similarities between character string and biostrings
```{r}
char.string <- "AATTGGAATTGGAAGGGG"
char.string
substr(char.string, start = 1, stop = 4) # Getting a substring from a character string

temp.biostring <- DNAString("AATTGGAATTGGAAGGGGN")
temp.biostring
temp.biostring[1:4] # Getting a substring from a biostrings

```

```{r}
# All characters that can be used in DNAStrings
DNA_ALPHABET

# All characters that can be used in AAStrings
AA_ALPHABET
```

### Useful functions to work with biostring
```{r, eval = F}
width() # Get the lengths of each DNAString in a DNAStringSet
reverseComplement() # Find the DNA sequence of the opposite strand
translate() # Convert DNA sequence to amino acid sequence
codons() # Split a DNAString sequence into codons
alphabetFrequency() # Find out the frequency of individual nucleotides in a StringSet
letterFrequency() # Find out the frequency of multiple nucleotides in a StringSet
dinucleotideFrequency() # Find out the frequency of dinucleotides in a StringSet
oligonucleotideFrequency() # Find out the frequency of any N-mer in a StringSet
```

### Lengths of DNAStrings
```{r}
width(yeast.fasta)
width(telomere.seq)
```

### Reverse complements of sequences
```{r}
telomere.seq
reverseComplement(telomere.seq)
```

### Convert DNA sequence to amino-acid sequence
```{r}
telomere.seq
translate(telomere.seq)

telomere.seq[[1]]
translate(telomere.seq[[1]])

# Translate in reading frame 2
translate(telomere.seq[[1]][2:width(telomere.seq)])
```

### Get codons from a sequence
```{r}
# Works only on DNAStrings not DNAStringSets
codons(telomere.seq[[1]])
```

### Frequency of all bases on each chromosome
```{r}
alphabetFrequency(yeast.fasta)
```

```{r}
alphabetFrequency(yeast.fasta[1])
alphabetFrequency(translate(yeast.fasta[1]))
```

### Get GC content of each chromosome
```{r}
# G OR C (GC-content)
letterFrequency(yeast.fasta, "GC") # GC counts
letterFrequency(yeast.fasta, "GC", as.prob = TRUE) # GC frequency
```

### Frequency of all pairs of nucleotides
```{r}
dinucleotideFrequency(yeast.fasta)
```

## Create CDS sequences of all genes without introns
```{r}
# Filter all CDS ranges from the GFF file
yeast.cds <- yeast.gff[yeast.gff$type == "CDS"]
yeast.cds
```

```{r}
# Get a list of genes without any introns
genes.wo.introns <- yeast.cds %>%
  as.data.frame() %>%
  group_by(Name) %>%
  tally() %>%
  filter(n == 1) %>%
  pull(Name)

yeast.cds.no.introns <- yeast.cds[yeast.cds$Name %in% genes.wo.introns]
yeast.cds.no.introns
```

```{r}
# Get the sequences of aCDS
yeast.cds.seq <- yeast.fasta[yeast.cds.no.introns]
yeast.cds.seq

# Add gene names to the new DNAStringSet
names(yeast.cds.seq) <- yeast.cds.no.introns$Name
yeast.cds.seq

# Make sure all sequences are in the right orientation
yeast.cds.seq[strand(yeast.cds.no.introns) == "-"] <- reverseComplement(yeast.cds.seq[strand(yeast.cds.no.introns) == "-"])
yeast.cds.seq
```

### Check if the CDS sequences begin with start and stop codons
```{r}
yeast.cds.start <- subseq(yeast.cds.seq, start = 1, end = 3)
yeast.cds.start
```
```{r}
oligonucleotideFrequency(yeast.cds.start, 3) %>%
  colSums()

table(yeast.cds.start)
```

```{r}
yeast.cds.stop <- subseq(yeast.cds.seq, 
                         start = width(yeast.cds.seq) - 2,
                         end = width(yeast.cds.seq))
yeast.cds.stop

table(yeast.cds.stop)
```

## Pattern matching
```{r}
cpg.pattern = "CGCGCGCGCG"
vcountPattern(pattern = cpg.pattern, 
              subject = yeast.fasta, 
              max.mismatch = 0)

vcountPattern(pattern = cpg.pattern, 
              subject = yeast.fasta, 
              max.mismatch = 1)
```

```{r}
all.perfect.matches <- vmatchPattern(pattern = cpg.pattern, 
              subject = yeast.fasta, 
              max.mismatch = 0)
all.perfect.matches[4]

mismatch.cpg <- vmatchPattern(pattern = cpg.pattern,
              subject = yeast.fasta,
              max.mismatch = 1)
mismatch.cpg[4]
```











