---
title: "Lab-03 in-class"
output: html_notebook
---

## Help
```{r}
help(seq)
?seq
```

## Exercise 2.4 (from previous class)
### 4. Subset a new 2x2 matrix from the above matrix with the following values
```{r}
num_mat_3x5 <- matrix(1:15, nrow = 3, byrow = TRUE) # Create a 3 by 5 matrix
num_mat_3x5

mat_right_rows <- num_mat_3x5[c(2, 3), ] # Subset rows first
mat_right_rows

mat_right_rows_cols <- mat_right_rows[, c(2, 4)] # From the previous matrix, subset columns now
mat_right_rows_cols

mat_right_rows_cols2 <- num_mat_3x5[c(2, 3), c(2, 4)] # Subset rows and columns simultaneuously
mat_right_rows_cols2
```

## DataFrames
```{r}
# Create a new data frame
my_data <- data.frame(
    Gene = c("Gene1", "Gene2", "Gene3", "Gene4"),
    Length = c(100,200,300, 400),
    Disease = c(TRUE, FALSE, FALSE, TRUE),
    stringsAsFactors = FALSE
)
my_data

# Each column is a vector and has a name and is separated by commas from other columns.
```

```{r}
# Length gives you number of columns
length(my_data)

# Dimensions of a data frame
dim(my_data)

# Number of rows of a data frame
nrow(my_data)

# Number of columns of a data frame
ncol(my_data)

# Structure of a data frame
str(my_data)

# Identify type of objects
class(my_data)
class(my_data$Disease)
class(my_data[, 3])

# Check the type of data in a column
is.logical(my_data$Disease)
is.numeric(my_data$Disease)
```

```{r}
# Convert the Disease column from a logical type to a character type
my_data$Disease <- as.character(my_data$Disease)
my_data

# Check the type of the new Disease column
is.logical(my_data$Disease)

```



## Exercise 3.1
### 1. Read in the data.tsv file from didact and load it as a data frame.
```{r}
my_data <- read.table("data.tsv", header = TRUE, stringsAsFactors = FALSE)
my_data
```

### 2. Find the number of genes in the file using three different R commands.
```{r}
# Method 1
length(my_data$Gene)

# Method 2
nrow(my_data)

# Method 3
dim(my_data)[1]

# Method 4
str(my_data)
```


### 3. Create a new data frame which contains only the first three columns of the old data frame. 
```{r}
new_data <- my_data[, 1:3]
new_data
```


### 4. What would the length of a genome be if it was made up only the genes listed in this file and no non-coding regions?
```{r}
sum(my_data$Length)

sum(my_data[, 2])

sum(my_data[, "Length"])
```


## Exercise 3.2
### 1. How many genes have uATGs in them?
```{r}
my_data
```

```{r}
# Method 1 - Use logical comparison and which statement to select all genes with uATG and get the length of the new vector
length(which(my_data$uATG == TRUE)) 

# Method 2 - Since the uATG column is a logical vector, use the which statement directly to select all genes with uATG and get the length of the new vector
length(which(my_data$uATG))

# Method - Since the uATG column is a logical vector, use the numeric value of TRUE/FALSE to count the number of genes with uATGs
sum(my_data$uATG)

```


### 2. How many genes have utrs?
```{r}
length(which(my_data$UTR != 0)) # While this works, it is not ideal
length(which(my_data$UTR > 0)) # While this works, it is not ideal

length(which(!is.na(my_data$UTR))) # Use is.na to create a logical vector of location of NA, TRUE is NA. Then use ! to negate this - so now NAs are FALSE and numeric values are TRUE. Identify the location of the new TRUE values and get their length.
```


### 3. Create a new data frame with only genes longer than 500 codons.
```{r}
var_pos_500 <- my_data$Length > 500 # Create a logical vector of all location of genes > 500
var_pos_500

which(var_pos_500) # Identifies the location of those genes

# Method 1
my_data[which(var_pos_500), ] # Use the identity of the locations to subset the dataframe

# Method 2
my_data[var_pos_500, ] # Use the logical vector directly to subset. All positions with TRUE will be selected and all positions with FALSE will be ignored.

```


### 4. Where in the dataset are the following genes - YBL006C, YBL025W, and YAL025C?
```{r}
sub_genes <- c("YBL006C", "YBL025W", "YAL025C") # Create a vector of genes of interest

my_data$Gene %in% sub_genes # Use %in% function get a logical vector of positions of these genes.

pos_sub_genes <- which(my_data$Gene %in% sub_genes) # Get the position/indices of these genes.
pos_sub_genes

my_data[pos_sub_genes, ] # Bonus - get the data associated with these genes.
```

### 5. What is the combined length of these 3 genes?
```{r}
new_data <- my_data[pos_sub_genes, ] # Save the data frame associated with the three genes in a new dataframe
new_data

sum(new_data$Length) # Sum the lengths of the three genes/
```













