---
title: "Lab 12 in-class"
output:
  html_document:
    df_print: paged
---

```{r}
library(GenomicRanges)
library(tidyverse)
```


```{r}
var.irange <- IRanges(start = c(4, 5, 6), end = c(11, 12, 13))
var.irange
```

```{r}
var.grange <- GRanges(seqnames = c("chr1", "chr2", "chr3"),
                      strand = c("+", "-", "+"),
                      ranges = var.irange)
var.grange
```

```{r}
values(var.grange) <- DataFrame(gname = c("gene1","gene2","gene3"),
                                type = c("enzyme","structural","enzyme"))
var.grange
```

```{r}
seqlevels(var.grange) <- c("chr1", "chr-mt", "chr2", "chr3")
seqlengths(var.grange) <- c("chr1" = 100, "chr-mt" = 200, "chr2" = 400, "chr3" = 500)
isCircular(var.grange) <- c(FALSE, TRUE, FALSE, FALSE)
genome(var.grange) <- "Mouse"
```

```{r}
var.grange
```

```{r}
seqinfo(var.grange)
```

Yeast GFF
```{r}
library(rtracklayer)

# Read in the yeast GFF file as GRanges. 
yeast.gr <- readGFFAsGRanges("saccharomyces_cerevisiae.gff.txt")
```
```{r}
yeast.gr <- readGFFAsGRanges("saccharomyces_cerevisiae.gff.txt")
yeast.gr
```



## Exercise 12
### 1. How many chromosomes are there in the yeast genome?
```{r}
yeast.gr %>%
  as.data.frame() %>%
  group_by(seqnames) %>%
  tally() %>%
  nrow()

seqinfo(yeast.gr) %>%
  length()

yeast.gr
```


### 2. How many genes encode proteins?
```{r}
table(yeast.gr$type)

yeast.gr %>%
  as.data.frame() %>%
  group_by(type) %>%
  tally()

yeast.gr %>%
  as.data.frame() %>%
  filter(type == "gene")

yeast.gr %>%
  as.data.frame() %>%
  filter(type == "mRNA") %>%
  nrow()
```
















