---
title: "Lab-16 Working with genome sequences - Biostrings"
author: "Premal Shah"
date: "Nov 04, 2021"
output: 
  html_notebook:
    theme: spacelab
    toc: yes
    toc_float: yes
---

## Load packages
```{r}
.libPaths('/projects/community/py-image/2019.10A/bd387/anaconda3-intel17/envs/shahclass/lib/R/library')
library(Biostrings)
library(rtracklayer)
library(tidyverse)
```

## Read in yeast annotation and genome sequences
```{r}
```

## Homework from Lab-15

1. There are 6600 protein-coding genes in the yeast genome. How many of these genes are on the positive strand versus negative strand? (2 pts)
2. What fraction of these genes are on the mitochondria? (2 pts)
3. Which chromosome has the most number of tRNAs? (2 pts)
4. Find the longest gene on chromosome II and get it's sequence. (4 pts)

Bonus question:
1. How many of the 6600 protein-coding genes have introns?


## Working with Biostrings
```{r}
width() # Get the lengths of each DNAString in a DNAStringSet
length() # The number of elements in a DNAStringSet

reverseComplement() # Find the DNA sequence of the opposite strand

# Get first 99 bases of a the first String

translate() # Convert DNA sequence to amino acid sequence
translate()

alphabetFrequency() # Find out the counts of individual nucleotides/AAs in a String/StringSet

letterFrequency() # Find out the counts of multiple nucleotides in a StringSet

dinucleotideFrequency() # Find out the frequency of dinucleotides in a StringSet
oligonucleotideFrequency() # Find out the frequency of any N-mer in a StringSet
```

## Exercise 16.1
1. Create a DNAStringSet of CDS of all intron-less genes. 

## Exercise 16.2
1. Which chromosome has the highest GC content?
2. Which genes have the most number of arginine amino acids and which ones have the highest proportion of arginine amino acids?
3. Lysine is encoded by AAA and AAG codons. Which gene has the highest proportion of AAG encoded lysines? Consider only genes with at least 20 lysine amino acids. Hint: read the help file of `oligonucleotideFrequency` carefully.

## Subsetting Biostrings
```{r, eval=FALSE}
subseq(DNAString/DNAStringSet, start = , end = , width = )
```

## Pattern Matching
- Count the number of times a pattern is found in Biostrings
```{r, eval=FALSE}
# For single DNAString/AAString
countPattern(PATTERN, SUBJECT, MISMATCH ARGS)

# For DNAStringSet/AAStringSet
vcountPattern(PATTERN, SUBJECT, MISMATCH ARGS)
```

- Use `data(HNF4alpha)`

## Pattern Matching
- Find locations of patterns in Biostrings
```{r, eval=FALSE}
# For single DNAString/AAString
matchPattern(PATTERN, SUBJECT, MISMATCH ARGS)

# For DNAStringSet/AAStringSet
vmatchPattern(PATTERN, SUBJECT, MISMATCH ARGS)
```

- Use `data(HNF4alpha)`

## Exercise 16.3
1. Create forward and reverse primers that are 30bp long for each CDS in the yeast genome. 
2. Which CDS in the yeast genome have long stretches (15 bp) of same nucleotide? Which bases are repeated more often? 
3. Which amino acids are most abundant in the C-terminal ends (last 5 residues) of yeast proteins?